### ISSUE TITLE
Strange behavior

### ISSUE THREAD

**Issue by conradoqg** [pago/pioc#3]
I'm trying to understand why the following situation does not works. Can you help me?

Example 1:

``` javascript
 var module = pioc.createModule()
    .bind('services/consoleService', require('./services/consoleService.js'))
    .bind('services/dummyService', require('./services/dummyService.js'))
...
injector.resolve(function (services) {
      console.log(util.inspect(services)); // Array[2] - Expected
});

injector.resolve(function (consoleService) {
      console.log(util.inspect(consoleService)); // Array[0] - Should not return the consoleService?
});
```

Example 2:

``` javascript
 var module = pioc.createModule()
    .bind('services_consoleService', require('./services/consoleService.js'))
    .bind('services_dummyService', require('./services/dummyService.js'))
...
injector.resolve(function (services) {
      console.log(util.inspect(services)); // Array[2] - Expected
});

injector.resolve(function (services_consoleService) {
      console.log(util.inspect(services_consoleService)); // Object - It works but it's not pretty
});
```

Thank you!

**Comment by pago** [comment]
The service name is resolved explicitly unless there is no such service. In that case, it'll assume that the service name is actually a prefix and will try to find all services that share this prefix.
So the examples you've shown are actually all exactly as I'd expect them.

I agree with you however, that the full name thing is ugly. Especially for constructor/function injection. pioc supports AngularJS style annotations, so you could (if you really wanted to) wrap the last case like this:

``` javascript
injector.resolve(['services_consoleService', function(consoleService) {
     // ...
}]);
```

But that is really only ever practical for client side use.

For my own code, I try to use property injection instead of constructor injection whenever possible as it doesn't have the issues around minification on the client side and is more explicit.

If you have a proposal for a better handling for constructor/function injection, I'm all ears. I'll invest some time as well to try to come up with a better solution.

**Comment by conradoqg** [comment]
I understand now, thank you.

What do you think of adding something like 'namespace' in service name? For example:

``` javascript
 var module = pioc.createModule()
    .bind('services/groupA/consoleService', require('./services/consoleService.js'))
    .bind('services/groupA/dummyService', require('./services/dummyService.js'))
    .bind('services/groupB/anotherService', require('./services/anotherService.js'))
    .bind('name/consoleService', require('./services/consoleService.js'))
...
injector.resolve(function (services) {
      console.log(util.inspect(services)); // Array[3] (console, dummy and another)
});

injector.resolve(function (groupA) {
      console.log(util.inspect(groupA)); // Array[2] - (console, dummy)
});

injector.resolve(function (consoleService) { 
      console.log(util.inspect(services_consoleService)); // Array[2] - (consoleService from services/, consoleService from name/)
});
```

However, I'm not sure if this is a good solution. 
The simple solution maybe is to allow to find service by name, ignoring the prefix, when the search by "prefix" result in nothing

**Comment by pago** [comment]
Sorry, was busy at work.
I believe that namespaces are too much, however, I agree that at least a suffix search should be performed before giving up.

In the end, the task of a DIC is to resolve dependencies and resolving to something should be the default behaviour (and in most cases should be preferable to failing). However, I'd like to add some configuration options to it so that the various fallback behaviours can be disabled.

**Comment by pago** [comment]
Behaviour has been implemented in v1.3.0 (just released).
